Synchronize SQL timestamps with KV timestamps.

knz opened the pull request:
User-visible changes:
- the SQL `transaction_timestamp()` is now following the KV txn  timestamp. This buys us monotonically increasing SQL timestamps (in the same order as txn commit order)
- `current_timestamp()`, `now()` and `transaction_timestamp()` now refer to the same thing. `age()` computes relative to that too. (This is also what PostgreSQL does).
- `statement_timestamp()` evolves with each statement executed, as in PG.
- a new SQL function `transaction_timestamp_unique()` enables clients to observe strictly monotonically increasing timestamps across different timestamps (when combined with `transaction_timestamp()`).

Fixes #4393

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5037)

<!-- Reviewable:end -->

tschottdorf commented:
I envisioned the clock passed into `Executor` and then a new `MinInitialTimestamp` field added to the options to `(*Txn).Exec`. This field would be populated in the executor. I don't see a reason to make the HLC available to the `client`. 

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-:-KCSZIRVrmQinNvlHp7M:1240042157)*
<!-- Sent from Reviewable.io -->

knz commented:
ok lemme try again

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-:-KCSZt0B8BKKXPEQmoRZ:-235333688)*
<!-- Sent from Reviewable.io -->

knz committed:
Synchronize SQL timestamps with KV timestamps.

User-visible changes:

- the SQL transaction_timestamp() is now following the KV txn
  timestamp.

- current_timestamp(), now() and transaction_timestamp() now refer to
  the same thing. age() computes relative to that too.
  (This is also what PostgreSQL does).

- statement_timestamp() evolves with each statement executed, as in PG.

- a new SQL function transaction_timestamp_unique() enables clients
  to observe strictly monotonically increasing timestamps across
  different timestamps (when combined with transaction_timestamp()).

tschottdorf commented:
There's a TODO on `Session_Transaction.TxnTimestamp` which should be addressed/removed.

The code appears to be pretty broken still. Since you removed the WIP label, I assumed it was ready - should readd that until it is (I'll still take another look when you ping me).

---

Reviewed 5 of 10 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 432 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTQTM4N2w3vapJPTsl:-KCTQTM4N2w3vapJPTsm:-1419649065) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/client/txn.go#L432)):*
s/Initial/initial/

---

*[client/txn.go, line 449 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTT6uWS8Ryb-C9p1Ci:-KCTT6uWS8Ryb-C9p1Cj:-598270394) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/client/txn.go#L449)):*
Would be good to see a simple (client level) test for this.

---

*[client/txn.go, line 456 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTQZ97pciBdKC56qlq:-KCTQZ97pciBdKC56qlr:-197853461) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/client/txn.go#L456)):*
why did this change?

---

*[client/txn.go, line 506 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTQmawwfnYqoV9oCHg:-KCTQmawwfnYqoV9oCHh:2037836653) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/client/txn.go#L506)):*
Seems unrelated and not obviously better, please explain.

---

*[sql/executor.go, line 474 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTS7QVs7jGmSthmL0H:-KCTS7QWQzNCwPgp40Sz:-247551957) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/sql/executor.go#L474)):*
no timestamp here?

---

*[sql/executor.go, line 584 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTSDZ9GiYgqsIN73PE:-KCTSDZ9GiYgqsIN73PF:-478643510) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/sql/executor.go#L584)):*
in the case in which we're not in a transaction but the first item **is** BeginTransaction, where do we chose the timestamp?

---

*[sql/executor.go, line 707 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTRaEnImJL57mkf1Zf:-KCTRaEoKx8muESA1G8V:1779944291) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/sql/executor.go#L707)):*
we should internally deal with the whole timestamp and not with `time.Time`, so that's what `setTxn` should receive.

---

*[sql/parser/builtins.go, line 554 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTSeisRKPkxq-MeVYH:-KCTSeisRKPkxq-MeVYI:37757261) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/sql/parser/builtins.go#L554)):*
I never see TxnTimestamp set in this diff. Why is that?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-:-KCS_PncqNPR8v6733Dz:-1585946899)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
the basic direction looks good though, btw.

knz commented:
Review status: 0 of 13 files reviewed at latest revision, 8 unresolved discussions.

---

*[client/txn.go, line 456 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTQZ97pciBdKC56qlq:-KCrOVFuEaPACNN9vgay:-1769132879) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/client/txn.go#L456)):*
After all it's legit to call Exec on a nil txn, that happens e.g. with ROLLBACK.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Looks good! A few minor comments, should be good to go soon.

---

Reviewed 5 of 10 files at r2, 13 of 13 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[client/txn.go, line 460 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhmZksWdtZaegLIMr:-KCuhmZksWdtZaegLIMs:-468236923) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/client/txn.go#L460)):*
Add a simple test for this in `client/txn_test.go` (or wherever it fits).

---

*[kv/txn_coord_sender.go, line 477 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhcTAaAtCqepT1zko:-KCuhcTAaAtCqepT1zkp:-572345206) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/kv/txn_coord_sender.go#L477)):*
add a simple test for this in (I think) `txn_coord_sender_test.go`.

---

*[sql/executor.go, line 302 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCueNvFOj6vCFfeZuA9:-KCueNvFOj6vCFfeZuAA:-298625324) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/executor.go#L302)):*
Could you add comments on the field definitions in `type EvalContext struct`?

---

*[sql/executor.go, line 764 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCufPVjZvztngWUu8EN:-KCufPVjZvztngWUu8EO:404837729) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/executor.go#L764)):*
there's something like this above in the diff (where it's set into EvalCtx). I thought the evalCtx would be in the txnState or somewhere accessible here.

---

*[sql/plan.go, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCufqlKd2nvx4xW3MHz:-KCufqlKd2nvx4xW3MI-:-409885508) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/plan.go#L66)):*
if txn is nil, we should set ZeroTimestamp.

---

*[sql/parser/builtins.go, line 554 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTSeisRKPkxq-MeVYH:-KCug5Gqjh4IhqtG9l62:-17302356) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/sql/parser/builtins.go#L554)):*
Converting to `GoTime()` loses information which could lead to a loss of linearizability (not in the case in which it's guaranteed through waits, but in the case where we usually have it due to causal ordering). No real way around this unless we expose HLC timestamps to the client, but good to keep in mind - maybe in a comment on EvalCtx.

---

*[sql/parser/builtins.go, line 563 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCugZKbpUE3AFdQi3T_:-KCugZKbpUE3AFdQi3Ta:-638792694) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/builtins.go#L563)):*
ah, this "addresses" the above comment. This works (I'm assuming returning the whole HLC timestamp here is more awkward to work with for clients).

---

*[sql/parser/eval.go, line 754 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuh28YM3n5Mll5W9xx:-KCuh28YM3n5Mll5W9xy:209472186) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L754)):*
I don't think you want `log.Fatal` (which I think calls `os.Exit(1)`, so it's not recoverable) - use `panic` for assertions like these.

---

*[sql/parser/eval.go, line 762 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhF0Iqd8kHxsEAITh:-KCuhF0Iqd8kHxsEAITi:2050275022) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L762)):*
if you have a getter, see if you can unexport `TxnTimestamp` (otherwise it's not obvious that you're not supposed to access the field directly).
Ditto for `StmtTimestamp`.

---

*[sql/parser/eval.go, line 763 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuh9oYTyIC2bu2vPxz:-KCuh9oZLGkPk6z2HqvK:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L763)):*
ditto

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-:-KCuhsRXJXHDP-LQY5tw:1486425332)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 11 of 15 files reviewed at latest revision, 10 unresolved discussions.

---

*[client/txn.go, line 460 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhmZksWdtZaegLIMr:-KCv6SvaMjU7XXb2AP5r:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/client/txn.go#L460)):*
Done.

---

*[kv/txn_coord_sender.go, line 477 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhcTAaAtCqepT1zko:-KCuyv59IQJ_ui2EIiES:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/kv/txn_coord_sender.go#L477)):*
Done.

---

*[sql/executor.go, line 302 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCueNvFOj6vCFfeZuA9:-KCuvVUDYc3q9mJJAnNl:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/executor.go#L302)):*
Done.

---

*[sql/executor.go, line 764 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCufPVjZvztngWUu8EN:-KCuqqd-vKJE-FQbafnt:-1606722080) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/executor.go#L764)):*
We're using a different planMaker for Prepare (what you saw above) and Execute (this one). Both need to intialiaze their evalCtx. Although now that I think of it Prepare doesn't need timestamps; there should be no reading of timestamps during Prepare (SQL functions are not evaluated).

---

*[sql/plan.go, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCufqlKd2nvx4xW3MHz:-KCurTnsxE26ZbIDMsyr:1806705020) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/plan.go#L66)):*
I fail to agree here. setTxn is only called once on a freshly made planner, and ZeroTimestamp is the default already.

---

*[sql/parser/builtins.go, line 554 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCTSeisRKPkxq-MeVYH:-KCurfSHx1g-OEFo1qXd:-103707798) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d35a45d1443a320f23e5af2488428ccb89165f8/sql/parser/builtins.go#L554)):*
It's more complicated than that. Filed as #5266.

---

*[sql/parser/builtins.go, line 563 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCugZKbpUE3AFdQi3T_:-KCuueuyns8YI4vWkuLQ:-1540968692) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/builtins.go#L563)):*
See #5266 too.

---

*[sql/parser/eval.go, line 754 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuh28YM3n5Mll5W9xx:-KCuuvFzmBmwCTUsUa7B:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L754)):*
Done.

---

*[sql/parser/eval.go, line 762 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhF0Iqd8kHxsEAITh:-KCuv8Du7xSAEaICs7kC:795804306) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L762)):*
The thing is there are still multiple locations where it's valid to access the Timestamp fields when they are 0. Refactoring this would be a non-trivial cleanup with only moderate relevance to the bug being fixed.

---

*[sql/parser/eval.go, line 763 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuh9oYTyIC2bu2vPxz:-KCuvS7u3JjJdC4UW05K:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L763)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[client/txn_test.go, line 596 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvKdV0I7ryArt-JqYz:-KCvKdV0I7ryArt-JqZ-:-419142087) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5e8564fd9560d6c015e6045cd50f2b9640b6bf/client/txn_test.go#L596)):*
nit: go style would prefer `refTimestamp`.

---

*[kv/txn_coord_sender_test.go, line 169 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvKk1PGxE6x_CJhkHv:-KCvKk1PGxE6x_CJhkHw:-1677837463) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5e8564fd9560d6c015e6045cd50f2b9640b6bf/kv/txn_coord_sender_test.go#L169)):*
ditto.

---

*[kv/txn_coord_sender_test.go, line 183 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvKuy_jpOjU82TM8OF:-KCvKuy_jpOjU82TM8OG:-1667066412) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5e8564fd9560d6c015e6045cd50f2b9640b6bf/kv/txn_coord_sender_test.go#L183)):*
you can s/TxnMeta\.// (the `TxnMeta` struct is embedded). Not required, just saying.

---

*[sql/plan.go, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCufqlKd2nvx4xW3MHz:-KCvKI054LcKSqreVX-3:120845457) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/plan.go#L66)):*
That may be true, but shouldn't the code here speak for itself? After all, what you're writing above isn't obvious to anyone who reads this.

---

*[sql/parser/eval.go, line 762 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhF0Iqd8kHxsEAITh:-KCvMUQQ_98enXE79Ysr:46643933) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L762)):*
Try cherry-picking `pr5037` (on our main repo), see if you like it. I think it makes sure that the right thing gets always called.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-:-KCvMjamsN14497oujwV:1417088721)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 12 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[client/txn_test.go, line 596 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvKdV0I7ryArt-JqYz:-KCvbcnCaOHfcAmccX3C:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5e8564fd9560d6c015e6045cd50f2b9640b6bf/client/txn_test.go#L596)):*
Done.

---

*[kv/txn_coord_sender_test.go, line 169 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvKk1PGxE6x_CJhkHv:-KCvbdfwjfIRjyk4WFTO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5e8564fd9560d6c015e6045cd50f2b9640b6bf/kv/txn_coord_sender_test.go#L169)):*
Done.

---

*[kv/txn_coord_sender_test.go, line 183 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvKuy_jpOjU82TM8OF:-KCvbi_cu-4YrfWXod8U:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5e8564fd9560d6c015e6045cd50f2b9640b6bf/kv/txn_coord_sender_test.go#L183)):*
Done.

---

*[sql/plan.go, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCufqlKd2nvx4xW3MHz:-KCvcrV6dcEYVlcyWh-C:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/plan.go#L66)):*
Done.

---

*[sql/parser/eval.go, line 762 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhF0Iqd8kHxsEAITh:-KCvd3E4ScPNBsrjl_A6:562844484) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L762)):*
Wait what? pr5037 is this one. Which PR are you referring to?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 12 of 15 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/plan.go, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCufqlKd2nvx4xW3MHz:-KCvhMumO7Czvo6H62U6:2007322898) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/plan.go#L66)):*
Despite the comment, much easier and less confusing to write `p.EvalCtx.Timestamp = roachpb.ZeroTimestamp`. This is a method that can set `txn` to nil but does not zero out the timestamp as would be intuitive, but instead posts a 7 line explanation about why that's ok to do (at least at the time of writing, assuming that none of the surrounding code changes). I'm confused as to why you find that preferable.

---

*[sql/plan.go, line 69 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvi3E52rqrLbrlgz0u:-KCvi3E52rqrLbrlgz0v:-585339495) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d9f692bc048627e5fc9593bcd38c2c6546b617b/sql/plan.go#L69)):*
s/tmap/tamp/

---

*[sql/parser/eval.go, line 762 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhF0Iqd8kHxsEAITh:-KCvh4_UlK8ZqpgHOcTa:-1059112861) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L762)):*
sorry, `git co pr5037`, I pushed a branch by that name.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 10 of 13 files at r4, 2 of 4 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/executor.go, line 475 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvhwQc_OyFPq7JHGWm:-KCvhwQdi7R-jQcVFYgI:-2139553491) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86dee62a3eae0ee56301e10112941cacb5fca7c/sql/executor.go#L475)):*
is this redundant with `(*Executor).newTxn`? when should one or the other be called?

---

*[sql/parser/eval.go, line 755 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCviXNRllqvB9Ttsq_U:-KCviXNRllqvB9Ttsq_V:1852014275) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d9f692bc048627e5fc9593bcd38c2c6546b617b/sql/parser/eval.go#L755)):*
nit: s/valid/nonzero/

---

*[sql/testdata/datetime, line 166 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvimQ6phZKAozsvyAl:-KCvimQ6phZKAozsvyAm:1484821685) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d9f692bc048627e5fc9593bcd38c2c6546b617b/sql/testdata/datetime#L166)):*
shopuld "nonsecond" be "nanosecond"?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 12 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/executor.go, line 475 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvhwQc_OyFPq7JHGWm:-KCvt9BPr0mJlF9M9awX:-2029281586) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86dee62a3eae0ee56301e10112941cacb5fca7c/sql/executor.go#L475)):*
newTxn is used when you need to start a txn from scratch and need to initialize it (including its Proto) from defaults. Here we don't need that because the Proto is loaded on the next line.

---

*[sql/plan.go, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCufqlKd2nvx4xW3MHz:-KCvuVQzlfddMlE3t1LR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/plan.go#L66)):*
Done.

---

*[sql/plan.go, line 69 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvi3E52rqrLbrlgz0u:-KCvsBYRrOKuu8s2ih7r:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d9f692bc048627e5fc9593bcd38c2c6546b617b/sql/plan.go#L69)):*
Done.

---

*[sql/parser/eval.go, line 762 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhF0Iqd8kHxsEAITh:-KCvtk8RGZTVDcR8k-I-:860848815) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L762)):*
I looked at pr5037 and this code is nowhere near sufficient :) I will have a look at it but the problem is hairier than it looks like IMHO.

---

*[sql/parser/eval.go, line 755 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCviXNRllqvB9Ttsq_U:-KCvsyUm1vqRtPVpgPbi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d9f692bc048627e5fc9593bcd38c2c6546b617b/sql/parser/eval.go#L755)):*
Done.

---

*[sql/testdata/datetime, line 166 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvimQ6phZKAozsvyAl:-KCvtwMbrYs5RbnC3qx0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d9f692bc048627e5fc9593bcd38c2c6546b617b/sql/testdata/datetime#L166)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/executor.go, line 475 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvhwQc_OyFPq7JHGWm:-KCvumJJf2UBUcshXu4s:-586690122) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86dee62a3eae0ee56301e10112941cacb5fca7c/sql/executor.go#L475)):*
OK, if there's a strong difference, fine by me.

---

*[sql/parser/eval.go, line 758 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvuycR-AugcHlHhSQR:-KCvuycR-AugcHlHhSQS:-56597605) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5e8564fd9560d6c015e6045cd50f2b9640b6bf/sql/parser/eval.go#L758)):*
sorry, missed this last time. s/invalid/zero/ and below

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 13 of 15 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/parser/eval.go, line 762 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCuhF0Iqd8kHxsEAITh:-KCw-Y_E2E9TA6dmeziO:1003099910) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f9d2b472f0bcaecf9cea7efa8fdc5be9f2aaa5/sql/parser/eval.go#L762)):*
Oh I see you only touch txnTimestamp and not StmtTimestamp. That's easier.

---

*[sql/parser/eval.go, line 758 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-KCvuycR-AugcHlHhSQR:-KCvzzUoAOl_kmIlnlbt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5e8564fd9560d6c015e6045cd50f2b9640b6bf/sql/parser/eval.go#L758)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, time to get this in as far as I'm concerned :)

---

Reviewed 1 of 3 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037#-:-KCzXofvN_CVlWrsL3Ld:1878910098)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5037)*
<!-- Sent from Reviewable.io -->

knz commented:
@tschottdorf is merging this conditional on fixing #3525?

tschottdorf commented:
No, merge away - this is strictly an improvement.

On Wed, Mar 16, 2016 at 11:38 AM kena <notifications@github.com> wrote:

> @tschottdorf <https://github.com/tschottdorf> is merging this conditional
> on fixing #3525 <https://github.com/cockroachdb/cockroach/issues/3525>?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/5037#issuecomment-197387241>
>
